src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs

```diff
@@ -0,0 +1,300 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.ML.Data;
+using Microsoft.ML.Internal.Utilities;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Transforms.TimeSeries
+{
+    public class SrCnnAnomalyDetectionBase : IStatefulTransformer, ICanSaveModel
+    {
+        /// <summary>
+        /// Whether a call to <see cref="ITransformer.GetRowToRowMapper(DataViewSchema)"/> should succeed, on an
+        /// appropriate schema.
+        /// </summary>
+        bool ITransformer.IsRowToRowMapper => ((ITransformer)InternalTransform).IsRowToRowMapper;
+
+        /// <summary>
+        /// Create a clone of the transformer. Used for taking the snapshot of the state.
+        /// </summary>
+        /// <returns></returns>
+        IStatefulTransformer IStatefulTransformer.Clone() => InternalTransform.Clone();
+
+        /// <summary>
+        /// Schema propagation for transformers.
+        /// Returns the output schema of the data, if the input schema is like the one provided.
+        /// </summary>
+        public DataViewSchema GetOutputSchema(DataViewSchema inputSchema) => InternalTransform.GetOutputSchema(inputSchema);
+
+        /// <summary>
+        /// Constructs a row-to-row mapper based on an input schema. If <see cref="ITransformer.IsRowToRowMapper"/>
+        /// is <c>false</c>, then an exception should be thrown. If the input schema is in any way
+        /// unsuitable for constructing the mapper, an exception should likewise be thrown.
+        /// </summary>
+        /// <param name="inputSchema">The input schema for which we should get the mapper.</param>
+        /// <returns>The row to row mapper.</returns>
+        IRowToRowMapper ITransformer.GetRowToRowMapper(DataViewSchema inputSchema)
+            => ((ITransformer)InternalTransform).GetRowToRowMapper(inputSchema);
+
+        /// <summary>
+        /// Same as <see cref="ITransformer.GetRowToRowMapper(DataViewSchema)"/> but also supports mechanism to save the state.
+        /// </summary>
+        /// <param name="inputSchema">The input schema for which we should get the mapper.</param>
+        /// <returns>The row to row mapper.</returns>
+        public IRowToRowMapper GetStatefulRowToRowMapper(DataViewSchema inputSchema)
+            => ((IStatefulTransformer)InternalTransform).GetStatefulRowToRowMapper(inputSchema);
+
+        /// <summary>
+        /// Take the data in, make transformations, output the data.
+        /// Note that <see cref="IDataView"/>'s are lazy, so no actual transformations happen here, just schema validation.
```

Comment:
It transforms the data by doing what? Adding new columns? If so, where could I find the way those new columns get computed?

Reply:
Added more details.